Cloudformation battery included template

danmason opened the pull request:
See #2896 

Adds a `cloudformation` directory with a number of templates to be used in conjunction to create a full stack on AWS for XTDB (S3 + Kafka XTDB nodes, setting up all the dependent infrastructure and running on ECS/EC2 instances, connectable via an application loadbalancer)

# Templates

There are a total of 5 templates, set up in roughly the following order:

## `xtdb-vpc` 

Sets up a VPC to be used by the other infrastructure.

**Setup infrastructure:**
- Two public subnets
- Two private subnets
- Internet gateway
- Public & private routing tables, route associations, etc
- Security group

**Outputs:**
- `VpcId`: ID of the VPC created by the stack
- `PublicSubnets`: IDs of the both of the public subnets created within the VPC
- `PrivateSubnets`: IDs of the both of the private subnet created within the VPC
- `SecurityGroupId`: ID of the security group created for the VPC

## `xtdb-s3` 

Doesn't depend on the VPC - sets up all S3 infrastructure to be used by XTDB: 

**Inputs:**
- `S3BucketName` - Desired name of the bucket which will contain the XTDB Object Store.

**Setup infrastructure:**
- An SNS topic that acts as a bucket changelog (used by the XTDB file watcher) 
- An S3 bucket for the object store that publishes changes to the SNS topic. 
- Relevant roles to allow the bucket to send notifications to SNS.
- A managed policy for all S3/SNS/SQS permissions.  

**Outputs:**
- `BucketName`: passing back out the same input parameter 
- `SNSTopicArn`: The ARN of the SNS Topic with the S3 notifications
- `S3AccessPolicyArn`:  ARN of the managed policy that has relevant SNS and S3 permissions, and ability to create SQS queues

## `xtdb-msk` (dependent on `xtdb-vpc`)

Sets up MSK for use as the XTDB log

**Inputs:**
- `VpcId`: ID of the VPC to host the MSK cluster on (from `xtdb-vpc`)
- `SecurityGroupId`: Group ID of the security group to be used by the ECS nodes (from `xtdb-vpc`)
- `MSKClusterName`: User desired name of the MSK kafka cluster used as the XTDB Kafka Log - defaults to `xtdb-cluster-kafka`
- `MSKVolumeSize`: The size in GiB of the EBS volume for the data drive on each broker node of the kafka cluster - defaults to `100`

**Setup infrastructure:**
- Two private subnets, a route table, route table associations, and a security group for the subnets to be used exclusively by MSK.
- An MSK cluster with relevant settings: 
  - Name set by `MSKClusterName` parameter.
  - Kafka version `2.6.2`  
  - Each broker running `kafka.t3.small`, with `MSKVolumeSize` of EBS storage
  - Setup to be used unauthenticated (though only accessible by other things on the VPC)
- A managed policy for all S3/SNS/SQS permissions.  

**Outputs:**
- `MSKClusterArn`: ARN of the MSK cluster created by the stack 
- `MSKClusterName`:  Name of the created MSK cluster
- `MSKAccessPolicyArn`:  ARN of the managed policy that has relevant MSK permissions

## `xtdb-alb` (dependent on `xtdb-vpc`)

Sets up application load balancer and all relevant infrastructure for that to route traffic across the ECS service

**Inputs:**
- `VpcId`: ID of the VPC to host the MSK cluster on (from `xtdb-vpc`)
- `SecurityGroupId`: Group ID of the security group to be used by the ECS nodes (from `xtdb-vpc`)
- `PublicSubnets`: List of public subnets to host the load balancer on (from `xtdb-vpc`)

**Outputs:**
- `TargetGroupArn`: ARN of the Target Group created for the nodes
- `LoadBalancerArn`: ARN of the Applicaiton Load Balancer created for the nodes
- `LoadBalancerUrl`: The Load-Balanced XTDB node URL - 'http://${ECSALB.DNSName}:3000'

## `xtdb-ecs` (dependent on `xtdb-vpc`,  `xtdb-msk`, `xtdb-s3` & `xtdb-alb`)

Sets up XTDB on ECS, using EC2 instances. 

**Inputs (grouped in the template):**
- Parameters/resources from other stacks:
  - `SecurityGroupId`: Group ID of the security group to be used by the ECS nodes (from `xtdb-vpc`)
  - `PublicSubnets`: List of public subnets to host the load balancer on (from `xtdb-vpc`)
  - `TargetGroupArn: ARN of the Target Group created for the nodes (from `xtdb-alb`)
  - `LoadBalancerArn: ARN of the  Application Load Balancer created for the nodes (from `xtdb-alb`)
  - `S3BucketName`: Name of the S3 bucket to use as the XTDB object store (from `xtdb-s3`)
  -  `SNSTopicArn`: The ARN of the SNS Topic which the S3 sends notifications to (from `xtdb-s3`)
  - `S3AccessPolicyArn`: ARN of the managed policy offering access to all the S3 permissions necessary for the object store (from `xtdb-s3`)
  - `MSKBootstrapServers Comma separated list containing all Kafka bootstrap server URLs from MSK (needs to be grabbed manually from the MSK cluster info)
  - MSKAccessPolicyArn: ARN of the managed policy offering access to all the MSK permissions (from `xtdb-msk`)
- ECS Configuration:
  - `ClusterName`: Name of the desired ECS cluster -  defaults to `xtdb-cluster`
  -  `EC2InstanceType`: EC2 Instance Type used for ECS Service - defaults to `i3.large` (storage optimized)
  - `DesiredCapacity`: Number of EC2 instances to launch in your ECS cluster / XTDB node tasks to run - defaults to `1`
  - `ImageId:` Used to grab an ECS Optimized Image from SSM Parameter Store - defalts to `/aws/service/ecs/optimized-ami/amazon-linux-2023/recommended/image_id` 

**Setup infrastructure:**
- Sets up an ECS cluster. 
- Sets up a cloudwatch log group for the ECS tasks.
- Sets up an IAM role to be used by the EC2 instances that has access to everything it needs:
    - Makes use of a number of ManagedPolicies, namely the ones from the `s3` stack, `msk` stack and the aws `AmazonEC2ContainerServiceforEC2Role`
- Creates a launch configuration to setup ECS enabled EC2 instances using the image provided from the parameters:
  - AMI id grabbed from SSM parameter store using the `ImageId` parameter.
  - Instance type of these set by parameter, `EC2InstanceType`
  - Has some user data needed to set up the instance as an ECS container, and to mount the NVME drives.  
- Sets up an autoscaling group to create the `DesiredCapacity` amount of EC2 instances.
- Sets up an ECS task for the XTDB node container: 
  - Expects to run on `EC2`, has the relevant capability.
  - NetworkMode: 'bridge' 
  - Uses the `xtdb-aws-ea` docker image.
  - Adds a `xtdb-storage` volume pointed to a location within the EC2's mounted NVME drive, then uses this as a mountpoint within the container definition.  
  - Has a `MemoryReservation` to set a minimum memory limit - should allow the container to use up to the amount of memory on the instanceit runs on.
  - Passes through the various environment variables expected by the node startup, from the cloudformation template:
    - `KAFKA_BOOTSTRAP_SERVERS` - the list of bootstrap server URLs from MSK
    - `XTDB_TOPIC_NAME` - set to xtdb-log
    - `XTDB_S3_BUCKET` - the name of the created S3 bucket.
    - `XTDB_SNS_TOPIC` - the ARN of the created SNS topic.
  - Exposes port 3000 (which the HTTP server runs on)
  - Outputs logs to the created log group. 
- Creates an ECS service:
  - Launch type set to EC2
  - Desired count set to the parameter `DesiredCapacity`
  - Load balancer set up on container port 3000 using the setup application load balancer/target group.
  - Uses the **PlacementConstraint** of `distinctInstance` to ensure one task is ran per EC2 instance. 
  
# Considerations

A few things that may be a bit weird/contentious and I think might require some thoughts/opinions:
- Networking - currently just make it publicly available via a load balancer. Is this desired?
- The ECS service setup, in particular:
  - `DesiredCapacity` being used to configure both how many instances we want, and how many tasks/nodes to run (since we have 1 task per instance) - might well want to rethink this.
  - Partially set up the above to allow ECS tasks to set their minimum memory reservation and then use as much memory as they have available - it **might** be the case we want to set memory hard limits and allow users to configure those (can then tune these however they please based on what EC2 instance image they run) 

# EDIT: Issues with the XTDB http client

I was testing it locally, trying to send some messages through the XTDB client with the load balancer URL (http://ecsalb-1380902940.eu-west-1.elb.amazonaws.com/) - ran into some issues, seems it's giving me back 400s.. Currently not entirely sure why, seems to be somehow related to the use of `hato` with the loadbalancer.

Body comes back `nil`, status `400`, and headers seem pretty strange:
```
{"connection" "keep-alive", "content-length" "54", "content-type" "text/html;charset=iso-8859-1", "date" "Tue, 28 Nov 2023 16:37:10 GMT", "server" "Jetty(11.0.18)"}
```

Wrote up a number of comments of various investigations into this issue - in the end, could see the following:
- When we connected via hato directly to ec2, it works fine.
- Calls to the ALB endpoint from the browser, from CURL, and from clj-http - all worked fine.
- When we connected via hato to the ALB endpoint:
   - It seems to get directed to the target group - ie, actually hits the server.
   - The server then seems to return a 400 - though never logs anything, indicating the server rejects the client request.
   - The loadbalancer then passes back out the 400 response.

Since the message hit the server, looked at Jetty's Debug logs:
- The message sent when connecting directly via ec2 passed successfully, performing a h2c upgrade, no HTTP2Settings header.
- The message sent from the loadbalancer failed, and caused jetty an exception, had a HTTP2Settings header it could not parse.

Explicitly setting the http client to `http-1.1` seemed to fix the issue - so a commit to do so was pushed to `2.x`. Will likely want to explicitly re-instate `http-2` at a later point.

danmason committed:
Initial cloudformation template including everything.

danmason committed:
Split out template to individual templates.

jarohen commented on cloudformation/xtdb_vpc.yml:
```diff
@@ -0,0 +1,88 @@
+      RouteTableId: !Ref PublicRouteTable
+      SubnetId: !Ref PublicSubnet2
+
+  SecurityGroup:
```
let's tighten this up if we can - assume we'll only want HTTPS from 0.0.0.0/0 and full access only _within_ the security group?

danmason replied:
:+1: Yep - that sounds much better, was being very permissive initially to get stuff working :)

danmason replied:
WIll similarly have to do a self-referential rule :)

jarohen commented on cloudformation/xtdb_msk.yml:
```diff
@@ -0,0 +1,161 @@
+        ToPort: 9092
+        SourceSecurityGroupId: !Ref SecurityGroupId
+  
+  MSKSelfIngressAllowRule:
```
does this need a separate resource, or can it be included under `MSKSecurityGroup` with the others?

danmason replied:
I think since it's self referential (ie, security group pointing to itself) it _has_ to be done separately: https://stackoverflow.com/questions/46962263/cloudformation-securitygroup-circular-reference

jarohen commented on cloudformation/xtdb_vpc.yml:
```diff
@@ -0,0 +1,88 @@
+      InternetGatewayId: !Ref InternetGateway
+      VpcId: !Ref VPC
+
+  PublicSubnet1:
```
`PublicSubnetOne`, for consistency with the MSK one?

jarohen commented on cloudformation/xtdb_s3.yml:
```diff
@@ -0,0 +1,100 @@
+Parameters:
+  S3BucketName:
+    Type: String
+    Default: xtdb-cluster-bucket
```
We can't really give a default for the XTDB bucket name, as these have to be globally unique - i.e. as soon as one user uses the default nobody else can

jarohen commented on cloudformation/xtdb_alb.yml:
```diff
@@ -0,0 +1,79 @@
+      SecurityGroups:
+        - !Ref SecurityGroupId
+
+  ECSTG:
```
`ECSTargetGroup`?

jarohen commented on cloudformation/xtdb_alb.yml:
```diff
@@ -0,0 +1,79 @@
+        - Type: forward
+          TargetGroupArn: !Ref ECSTG
+      LoadBalancerArn: !Ref ECSALB
+      Port: '3000'
```
this the external port? if so, 443?

danmason replied:
As discussed - port 80 & HTTP for now, will add instructions for setting up HTTPs

Steps would be, in essence:
- Set up SSL certificate in ACM
- Update vpc security group to allow traffic on port 443
- Update ALB exposed port to 443, protocol to HTTPs, add certificate

jarohen commented on cloudformation/xtdb_msk.yml:
```diff
@@ -0,0 +1,161 @@
+    Description: The size in GiB of the EBS volume for the data drive on each broker node of the kafka cluster.
+
+Resources:
+  PrivateSubnetMSKOne:
```
For MSK you create the subnets within the MSK CF template, but for the ECS ALB you have them passed in - worth picking the same strategy for both? if so, any preferences?

danmason replied:
Hmmm.. It's a fair question. On the one hand, I think I'd prefer to keep them closer with the stuff that's actually using them, so I'm happy to create them within the ALB  template, will need to also expose them from there, too.

On the other... It would really require a LOT of bits moving over from the vpc template, so maybe its better to keep all the VPC related setup together and expect bits to be passed to MSK (I'm okay with passing around lists of subnets and the like - wherever we create them we're going to have to pass them as a parameter to `xtdb-ecs` whatever happens so may as well make a pattern of it)

danmason replied:
Decided to go with the approach of creating all the subnets in the VPC template and passing references around :)

jarohen reviewed (commented):
cheers @danmason :pray:

jarohen commented on cloudformation/xtdb_ecs.yml:
```diff
@@ -0,0 +1,239 @@
+            Action: 'sts:AssumeRole'
+      ManagedPolicyArns:
+        - 'arn:aws:iam::aws:policy/service-role/AmazonEC2ContainerServiceforEC2Role'
+        - !Ref S3AccessPolicyArn
```
can/should these go on the ECS service role rather than the machine itself?

danmason replied:
I don't think they belong in the ECS service role - my understanding is that is more for the ECS service to register container instances. I did try placing them in the `TaskExecutionRole` - but that didn't work, it says the following when attempting to set up s3:

```
"User: arn:aws:sts::<accountid>:assumed-role/xtdb-stack-ecs-EC2InstanceRole-cq4KLVh142Ez/i-0221dfee3da46d752 is not authorized to perform: sqs:createqueue on resource: arn:aws:sqs:eu-west-1:<account-id>:xtdb-object-store-notifs-queue-7de3c104-82b7-4463-a224-d93f831776d7 because no identity-based policy allows the sqs:createqueue action (Service: Sqs, Status Code: 403, Request ID: fa4ced72-32cf-51b5-9c7f-b0923e8b76df, Extended Request ID: null)"}}
```

Seems to use the instance role for these things, so planning to leave as is?

jarohen replied:
that's weird, I could have sworn AWS gave you a way to give permissions to the individual service/task - they're normally pretty fine grained when it comes to these things, I doubt they'd want to give any service running on that machine the same permissions, and it'd also mean that the permissions of the service depended on the machine it happens to run on?

jarohen replied:
just to clarify, IIUC there's two roles you can put on the task-def:
1. [execution role](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_execution_IAM_role.html), as you say - this is for setting up the container itself, pulling the image, writing logs, etc - i.e. I think this was the wrong one?
2. [task role](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-iam-roles.html) - this is the one I thought was for giving permissions for the container to access other services (e.g. S3, SNS, SQS). `TaskRoleArn` on the CFN template.

That article does also [mention](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-iam-roles.html#task-iam-role-considerations) to be careful when running containers on EC2 - that the containers inherit the permissions of the box itself.

danmason replied:
Ahhhh - that was it! Got you - yeah, does sound like a **task role** is the way forward. I'll look about bringing that in and testing all reboots and works correctly :)

jarohen commented on cloudformation/xtdb_ecs.yml:
```diff
@@ -0,0 +1,239 @@
+    Default: xtdb-cluster
+
+  EC2InstanceType:
+    Description: EC2 Instance Type used for ECS Service (We recommend using one that is storage optimized)
```
worth confirming that Docker can use the instance storage, rather than just the (much smaller) root partition of the machine

danmason replied:
Spent a fair bit of time on this! Pushed a commit that ensures the NVME drive gets mounted at `/mnt/nvme-storage` (this doesnt happen automatically, needed to update the launch configuration), and then used that as a volume mount on the task definition/container. 

Had a little play around to be able to SSH onto the node - can confirm the drive seems to get mounted and the volume seems to get made, so pretty happy it _should_ be using the correct storage now :)

danmason committed:
Quick fixes based on PR feedback.

- ECSTG -> ECSTargetGroup
- PublicSubnet1 -> PublicSubnetOne
- PublicSubnet2 -> PublicSubnetTwo
- Remove default s3 bucket name (need to be globally unique).

danmason committed:
Move private subnet creation to VPC, pass in to MSK template.

danmason committed:
Restrict ingress traffic on security group, expose ALB on port 80.

danmason committed:
Mount NVME drives on startup, use as volume/mountpoint for task.

danmason committed:
Move s3/msk access to ECS Task Role / out of instance role.
